x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -520,6 +534,78 @@ export class SessionIndex {
     }
   }
 
+  /**
+   * Checks whether specific session is within a concurrent sessions limit.
+   * @param sessionValue Session index value to check against concurrent sessions limit.
+   */
+  async isWithinConcurrentSessionLimit(sessionValue: Readonly<SessionIndexValue>) {
+    const maxConcurrentSessions = this.options.config.session.concurrentSessions?.maxSessions;
+    if (maxConcurrentSessions == null || !sessionValue.usernameHash) {
+      return true;
+    }
+
+    let sessionsOutsideLimit: Array<SearchHit<SessionIndexValue>>;
+    try {
+      const searchResponse = await this.options.elasticsearchClient.search<SessionIndexValue>({
+        index: this.aliasName,
+
+        // Find all sessions created for the same user by the same authentication provider.
+        query: {
+          bool: {
+            must: [
+              { term: { usernameHash: sessionValue.usernameHash } },
+              { term: { 'provider.type': sessionValue.provider.type } },
+              { term: { 'provider.name': sessionValue.provider.name } },
+            ],
+          },
+        },
+
+        // Sort sessions by creation date in descending order to get the most recent session that's also outside the
+        // limit. This query relies on a default value for `missing` sort parameter which is `_last`, meaning that
+        // sessions without `createdAt` field ("legacy" sessions) are always considered older than the ones that have
+        // this field populated. For example, if the limit is 2 the resulting set might look like this:
+        // { createdAt: 3 } <-- the most recent session (within the limit, not returned because of `from`)
+        // { createdAt: 2 } <-- the second most recent session (within the limit, not returned because of `from`)
+        // { createdAt: 1 } <-- the third most recent session (outside the limit, *returned*)
+        // { createdAt: undefined } <--- the oldest "legacy" session (outside the limit, not returned because of `size`)
+        sort: [{ createdAt: { order: 'desc' } }],
+
+        // Improve performance by fetching just one field of one outside-the-limit session and not tracking total hits.
```

Comment:
__note:__ since this method is going to be in the hot path if the limit is configured I tried to optimize it as much as I could.